### PR TITLE
[linux-port] Correct conditional groupings

### PR DESCRIPTION
**Pull request by pow2clk**
Clang and gcc complain when AND and OR logical operations are both
used in a single conditional because of the possibility of mistakes
when depending solely on operator precedence. In fact, this change
fixes a few clear errors mostly in asserts that would never fire.
Fixes 13 clang and 22 gcc warnings.

Additionally, when two if statements are followed by an else, there
may be some ambiguity in terms of which if the else applies to. It
is the closest one, the second in the case of two, but it is another
area of potential mistakes. Adding braces around the code intended
as the block of the first if conditional clarifies this.
fixed 8 clang and 8 gcc warnings.

Clang complains sometimes when assignments take place in a
conditional without additional parenthesis around it. It can
indicate when what was meant to be an equal comparison is replaced
with an assignment.
fixes 2 clang warnings.

Contributes to https://github.com/google/DirectXShaderCompiler/issues/206

### CODE REVIEW

File: lib/DxcSupport/dxcapi.use.cpp [hunk 1]
```diff
@@ -19,7 +19,7 @@ namespace dxc {
 static void TrimEOL(_Inout_z_ char *pMsg) {
   char *pEnd = pMsg + strlen(pMsg);
   --pEnd;
-  while (pEnd > pMsg && *pEnd == '\r' || *pEnd == '\n') {
+  while (pEnd > pMsg && (*pEnd == '\r' || *pEnd == '\n')) {
```
Comment (pow2clk) [inline lib/DxcSupport/dxcapi.use.cpp]: Note change of functionality. Previously if the character in pMsg[0] was '\n', this loop would advance past the front of the array. Extremely unlikely on Windows. Still really unlikely on Linux, but not impossible.

File: lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp [hunk 2]
```diff
@@ -2021,8 +2021,8 @@ static Value *LoadVectorOrStructArray(ArrayType *AT, ArrayRef<Value *> NewElts,
       Value *EltVal = LoadVectorOrStructArray(EltAT, NewElts, idxList, Builder);
       retVal = Builder.CreateInsertValue(retVal, EltVal, i);
     } else {
-      assert(EltTy->isVectorTy() ||
-             EltTy->isStructTy() && "must be a vector or struct type");
+      assert((EltTy->isVectorTy() ||
+              EltTy->isStructTy()) && "must be a vector or struct type");
```
Comment (pow2clk) [inline lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp]: The location of the parens  don't actually matter here because the string is always true, but this better approximates the intent.

File: tools/clang/lib/AST/DeclPrinter.cpp [hunk 3]
```diff
@@ -330,9 +330,8 @@ void DeclPrinter::VisitDeclContext(DeclContext *DC, bool Indent) {
       continue;
     }
 
-    if (isa<AccessSpecDecl>(*D)
-        && !Policy.LangOpts.HLSL // HLSL Change - no access specifier for hlsl.
-        ) {
+    if (isa<AccessSpecDecl>(*D))
+      if (!Policy.LangOpts.HLSL) { // HLSL Change - no access specifier for hlsl.
```
Comment (ehsannas) [inline tools/clang/lib/AST/DeclPrinter.cpp]: I'm curious why this change was necessary.
Is there anything wrong with doing

```cpp
if (isa<AccessSpecDecl>(*D) && !Policy.LangOpts.HLSL)
```
?

File: lib/DxcSupport/dxcapi.use.cpp (hunk 1 above)
Comment (ehsannas) [reply lib/DxcSupport/dxcapi.use.cpp]: :+1:

File: lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp (hunk 2 above)
Comment (ehsannas) [reply lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp]: :+1:

File: tools/clang/lib/AST/DeclPrinter.cpp (hunk 3 above)
Comment (pow2clk) [reply tools/clang/lib/AST/DeclPrinter.cpp]: This was clang being. . . let's say extra cautious and maybe too clever by half. Clang realizes that Policy.LangOpts.HLSL is always true in this build. Consequently the overall expression always fails regardless of the return value of isa<AccessSpecDecl>() So the return value doesn't matter so it's kind of ignored, which there is an attribute added to it to make sure never happens.

The attribute is surely intended to make sure that it isn't called as a standalone function because its return value is its only effect and such a usage would be a problem. I think the clang warning is useful, this is a weird case where it is oversensitive, but it was easier to humor it than any alternative.

Comment (ehsannas) [reply tools/clang/lib/AST/DeclPrinter.cpp]: I see. Thanks for clarifying.
